src/http.rs

```diff
@@ -811,4 +1259,483 @@ mod tests {
+                        "grant_type=client_credentials&client_id=some_client_secret&client_secret=some_secret",
+                        request_body,
+                    );
+
+                    let token_valid_for_seconds: u32 = 5;
+                    let random_token = Alphanumeric.sample_string(&mut rand::thread_rng(), 16);
+                    let token = format!(r#"
+                    {{
+                        "access_token": "{}",
+                        "token_type": "bearer",
+                        "expires_in": {},
+                        "scope": "some-scope"
+                    }}
+                    "#, random_token, token_valid_for_seconds);
+                    Ok::<Response<Body>, hyper::Error>(Response::new(Body::from(token)))
+                }));
+            futures_util::future::ok::<_, Infallible>(svc)
+        });
+
+        tokio::spawn(async move {
+            Server::bind(&addr).serve(make_svc).await.unwrap();
+        });
+
+        // Wait for the server to start.
+        tokio::time::sleep(Duration::from_millis(10)).await;
+
+        // Simplest possible configuration for oauth's client connector.
+        let tls: vector_lib::tls::MaybeTls<(), TlsSettings> =
+            MaybeTlsSettings::from_config(&None, false).unwrap();
+        let proxy_connector = build_proxy_connector(tls, &ProxyConfig::default()).unwrap();
+        let auth_client = Client::builder().build(proxy_connector);
+
+        let token_endpoint = format!("http://{}", addr);
+        let client_id = String::from("some_client_secret");
+        let client_secret = Some(SensitiveString::from(String::from("some_secret")));
+        // Fake oauth server returns token which expires in 5sec, with grace period equal 2 seconds
+        // we will have each token cached for next 3seconds (after this time token will be treat as expired).
+        let two_seconds_grace_period: u32 = 2;
+
+        // That can looks tricky for the first time, but idea is simple, we mock get_now_fn,
+        // which is used internally by OAuth2Extension to decidy whether token is eligible for refreshing.
+        // In real life Duration since epoch in seconds, can be for example 1730460289 (November 1, 2024),
+        // but to simplify understanding we wil start with 11 seconds sicne epoch, and can progress.
+        // Each value (index) in vec, means invocation of get_now_fn by OAuth2Extension, so
+        // first call returns Duration::from_secs(11), second, Duration::from_secs(12) and so on,
+        // because of that we have full controll over time here.
+        let mocked_seconds_since_epoch = [11, 12, 20, 21, 22, 23];
+        let counter = Arc::new(Mutex::new(0));
+        let get_now_fn = move || {
+            let counter = Arc::clone(&counter);
+            let mut counter = counter.lock().unwrap();
+            let i = *counter;
+            *counter += 1;
+            Duration::from_secs(mocked_seconds_since_epoch[i])
+        };
```

Comment:
I think we can use tokio test utils instead of manual implementing this: 
* https://docs.rs/tokio/latest/tokio/time/fn.advance.html 
* `#[tokio::test(start_paused = true)]`
* When you do the above, the test implementation will use a fake clock but the actual code will use the real system clock.

Reply:
but how to do this when single call 
```
let second_acquisition = extension.get_token().await.unwrap();
```
is expected to have different results of same get_now() call, so I need to have n, n+1, in Java world I would mock it like this
```
when(time.getNow()).
   thenReturn(N)
   thenReturn(N+1)
   thenReturn(N+2)
   thenReturn(N+3)
```
here, I can only do advance, that's why I extracted this to function (which I can mock in tests)

Reply:
I see. Can you do something like this?

```
#[tokio::test(start_paused = true)]
async fn test_token_acquisition() {
    let result = extension.get_token().await.unwrap();
    time::advance(Duration::from_secs(1)).await;
    // fake time advanced by 1 second

    let result = extension.get_token().await.unwrap();
    time::advance(Duration::from_secs(1)).await;
    // fake time advanced by 1 second
    
    // repeat... 
}
```

Reply:
Let me explain a little bit clearer, under single call to method we want to test, so
```rust
extension.get_token().await.unwrap();
```
we have two calls to method we are mocking
```rust
SystemTime::now()
```
this is used, in two places (under the hood of this single get_token() call), first
```rust
if self.get_time_now().as_millis() < token.expires_after_ms {
```
and then
```rust
let token_will_expire_after_ms =
            OAuth2Extension::calculate_valid_until(now, self.grace_period, &token);
```

Reply:
I see, thank you for the explanation.  Just to be clear I am not against what you have here now. But I would love to refactor this so it's a bit cleaner.I think the problem is that the current implementation of `get_token` does too many things. 

What if we split it into three functions:
```
fn is_token_expired(&self, token: &Token) -> bool { /* ... */ }
fn get_cached_token() -> Option<ExpirableToken> { /* ... */ }
fn async fn request_token(
        &self,
    ) -> Result<ExpirableToken, Box<dyn std::error::Error + Send + Sync>> { /* ... */ }
```

It will make the tests a bit more verbose but the actual code only uses it once. So it won't be too verbose.

Reply:
```rust
    async fn get_token(&self) -> Result<String, vector_lib::Error> {
        if let Some(token) = self.acquire_token_from_cache() {
            return Ok(token.access_token);
        }

        //no valid token in cache (or no token at all)
        let new_token = self.request_token().await?;
        let token_to_return = new_token.access_token.clone();
        self.save_into_cache(new_token);

        Ok(token_to_return)
    }
```
isn't it split already ?

Reply:
method responsible for request itself is ... let's say big
```rust
    async fn request_token(
        &self,
    ) -> Result<ExpirableToken, Box<dyn std::error::Error + Send + Sync>> {
        let mut request_body =
            format!("grant_type=client_credentials&client_id={}", self.client_id);

        // in case of oauth2 with mTLS (https://datatracker.ietf.org/doc/html/rfc8705) we only pass client_id,
        // so secret can be considered as optional.
        if let Some(client_secret) = &self.client_secret {
            let secret_param = format!("&client_secret={}", client_secret.inner());
            request_body.push_str(&secret_param);
        }

        let builder = Request::post(self.token_endpoint.clone());
        let builder = builder.header("Content-Type", "application/x-www-form-urlencoded");
        let request = builder
            .body(Body::from(request_body))
            .expect("Error creating request");

        let before = std::time::Instant::now();
        let response_result = self.client.request(request).await;
        let roundtrip = before.elapsed();

        let response = response_result.inspect_err(|error| {
            emit!(http_client::GotHttpWarning { error, roundtrip });
        })?;

        emit!(http_client::GotHttpResponse {
            response: &response,
            roundtrip
        });

        if !response.status().is_success() {
            let body_bytes = hyper::body::aggregate(response).await?;
            let body_str = std::str::from_utf8(body_bytes.chunk())?.to_string();
            return Err(Box::new(AcquireTokenError { message: body_str }));
        }

        let body = hyper::body::aggregate(response).await?;
        let token: Token = serde_json::from_reader(body.reader())?;

        let now = self.get_time_now();
        let token_will_expire_after_ms =
            OAuth2Extension::calculate_valid_until(now, self.grace_period, &token);

        Ok(ExpirableToken {
            access_token: token.access_token,
            expires_after_ms: token_will_expire_after_ms,
        })
    }
```
but comparable to 

```rust
    pub fn send(
        &self,
        mut request: Request<B>,
    ) -> BoxFuture<'static, Result<http::Response<Body>, HttpError>> {
        let span = tracing::info_span!("http");
        let _enter = span.enter();

        default_request_headers(&mut request, &self.user_agent);
        self.maybe_add_proxy_headers(&mut request);

        let client = self.client.clone();
        let auth_extension = self.auth_extension.clone();

        let fut = async move {
            if let Some(auth_extension) = auth_extension {
                let auth_span = tracing::info_span!("auth_extension");
                auth_extension
                    .modify_request(&mut request)
                    .instrument(auth_span.clone().or_current())
                    .await
                    .inspect_err(|error| {
                        // Emit the error into the internal events system.
                        emit!(http_client::AuthExtensionError { error });
                    })
                    .context(AuthenticationExtensionSnafu)?;
            }

            emit!(http_client::AboutToSendHttpRequest { request: &request });
            let response: client::ResponseFuture = client.request(request);

            // Capture the time right before we issue the request.
            // Request doesn't start the processing until we start polling it.
            let before = std::time::Instant::now();

            // Send request and wait for the result.
            let response_result = response.await;

            // Compute the roundtrip time it took to send the request and get
            // the response or error.
            let roundtrip = before.elapsed();

            // Handle the errors and extract the response.
            let response = response_result
                .inspect_err(|error| {
                    // Emit the error into the internal events system.
                    emit!(http_client::GotHttpWarning { error, roundtrip });
                })
                .context(CallRequestSnafu)?;

            // Emit the response into the internal events system.
            emit!(http_client::GotHttpResponse {
                response: &response,
                roundtrip
            });
            Ok(response)
        }
        .instrument(span.clone().or_current());

        Box::pin(fut)
    }
```

Reply:
> isn't it split already ?

No, what I meant was to change the following:
```
let token = self.get_token().await?;
```
to:
```
let token = match self.get_cached_token() {
    Some(cached_token) if !self.is_token_expired(&cached_token) => {
        cached_token // Return the valid cached token
    }
    _ => {
        // No valid token found, refresh
        self.refresh().await?
    }
};
```

However I understand this is getting too complex and the code looks correct to me. 

1. You can do the above refactoring
2. Delete the tests that depends on fake time  and also delete `get_time_now_fn` from `OAuth2Extension`

Reply:
honestly I cannot understand benefits of this move, 
```rust
let token = match self.get_cached_token() {
    Some(cached_token) if !self.is_token_expired(&cached_token) => {
        cached_token // Return the valid cached token
    }
    _ => {
        // No valid token found, refresh
        self.refresh().await?
    }
};
```
this is the same amount of logic, moved one step lower in the invocation tree.
```
Delete the tests that depends on fake time and also delete get_time_now_fn from OAuth2Extension
```
I can leave this test, there will be just no part of the test for token refreshing mechanism (so, the same scope as with mtls test)

Reply:
> there will be just no part of the test for token refreshing mechanism (so, the same scope as with mtls test - because get_time_now_fn will be removed), is that ok ?

Yes that's absolutely fine 👍

Reply:
fixed.